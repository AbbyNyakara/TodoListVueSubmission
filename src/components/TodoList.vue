<template>
  <div class="todo_list">
    <div class="input__section">
      <input type="text" placeholder="Create a new todo" class="todo_input"
      v-model="newTodo" @keyup.enter="addTodo">
    </div>
    
    <div class="todo_items">
      <TodoItem v-for="(todo, index) in filteredTodos" :key="todo.id" class="todo_item"
       :todo="todo" :index="index" @deletedTodo="deleteTodo" @markCompleted="markCompleted">
      </TodoItem>

      <div class="footer_container">
        <div class="items_left">
          <div>{{itemsLeft}} items left</div>
        </div>

        <div class="filters">
          <p :class="{ active : filter == 'all' }"  @click="filter='all'">All</p>
          <p  :class="{ active : filter == 'active' }"  @click="filter='active'">Active</p>
          <p  :class="{ active : filter == 'completed' }"  @click="filter='completed'">Completed</p>
        </div>

        <div class="clear_completed">
          <button @click="deleteCompleted" class="clear_btn" >Clear Completed</button>
        </div>
      </div>
    </div>
    
  </div>
</template>

<script>
  import TodoItem from './TodoItem.vue'

  export default {
    name: "TodoList",
    components: {
      TodoItem
    },
    data () {
      return {
        newTodo: "",
        idForTodo: 5,
        filter: "all",
        todos: [
          {
            item: "Complete online JavaScript course",
            completed: false,
            id: 1,
          },
          {
            item: "Jog around the park 3x",
            completed: false,
            id: 2,
          },
          {
            item: "10 minutes meditation",
            completed: false,
            id: 3,
          },
          {
            item: "Read for 1 hour",
            completed: false,
            id: 4,
          }
        ]
      }
    },
    methods: {
      addTodo() {
        if (this.newTodo.trim().length == 0) {
          alert("Todo Item cannot be empty")
          return
        }

        this.todos.push({
          id: this.idForTodo,
          item: this.newTodo,
          completed: false
        })

        this.newTodo = ""
        this.idForTodo++
      },

      deleteTodo(index) {
        this.todos.splice(index, 1)
      },
      deleteCompleted() {
        this.todos = this.todos.filter(todo => !todo.completed)
      }, 
      markCompleted(data) {
        this.todos.splice(data.index, 1, data.todo)
      }
    },
    computed: {
      itemsLeft() {
        return this.todos.filter(todo => !todo.completed).length
      },
      filteredTodos() {
        if (this.filter == "all") {
          return this.todos
        } else if (this.filter == "active"){
          return this.todos.filter(todo => !todo.completed)
        }else if (this.filter == "completed"){
          return this.todos.filter(todo => todo.completed)
        }

        return this.todos
      }
    }
  }
</script>

<style scoped>
  .todo_list {
    margin-top: 2rem;
  }

  .todo_input {
    padding: 20px 10px;
    width: 100%;
    font-size: 1rem;
    background-color: hsl(235, 24%, 19%);
    border: none;
    outline: none;
    color: hsl(234, 39%, 85%);
    border-radius: 5px;
  }

  .todo_items {
    margin-top: 2rem;
    background-color: hsl(235, 24%, 19%);
    color: hsl(234, 39%, 85%);
    border: none;
    border-radius: 5px;
  }

  .todo_input::before {
    content: "";
    background: url("../assets/images/icon_circle.png");
    position: absolute;
    width: 20px;
    height: 20px;
    background-position: center;
    background-size: contain;
  }

  .footer_container {
    font-size: 0.9rem;
    padding: 10px;
    display: flex;
    justify-content: space-between;
  }

  .filters {
    display: flex;
  }

  .filters> p {
    cursor: pointer; 
    font-size: 14px;
    background-color: hsl(235, 24%, 19%);
    appearance: none;
    padding: 5px;
    border: none;
    outline: none;
  }

  .active {
    color: hsl(192, 100%, 67%);
  }

  .clear_btn {
    border: none;
    outline: none;
    background: transparent;
    color: #fff;
    cursor: pointer;
  }

</style>