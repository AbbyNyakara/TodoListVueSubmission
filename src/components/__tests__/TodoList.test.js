import { mount } from "@vue/test-utils";
import { expect, test, describe, it } from 'vitest'
import TodoList from '../../components/TodoList.vue'
import Vue from 'vue';

describe("My Todolist component", ()=> {
  it('Should render component', ()=> {
    const wrapper = mount(TodoList);
    expect(wrapper.html()).toMatchSnapshot();
  })

  // it ("Should delete a todo item from array", ()=> {
  //   function deleteTodo(index){
  //     array.splice(index, 1)
  //   }

  //   const array = [
  //     {
  //     id: 1,
  //     item: 'complete laundry',
  //     completed: false
  //     }
  //   ]
  //     // const wrapper = mount(TodoList);
  //      array.deleteTodo(1);
  //     expect(array.length).toBe(0);
  // })
})


