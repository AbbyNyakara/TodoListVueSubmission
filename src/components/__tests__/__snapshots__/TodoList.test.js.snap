// Vitest Snapshot v1

exports[`My Todolist component > Should render component 1`] = `
"<div class=\\"todo_list\\" data-v-4afa86a5=\\"\\">
  <div class=\\"input__section\\" data-v-4afa86a5=\\"\\"><input type=\\"text\\" placeholder=\\"Create a new todo\\" class=\\"todo_input\\" data-v-4afa86a5=\\"\\"></div>
  <div class=\\"todo_items\\" data-v-4afa86a5=\\"\\">
    <div class=\\"todo_item\\" data-v-971ec5b0=\\"\\" data-v-4afa86a5=\\"\\">
      <div class=\\"todo_label\\" data-v-971ec5b0=\\"\\"><input type=\\"checkbox\\" class=\\"checkbox\\" data-v-971ec5b0=\\"\\">
        <div class=\\"todo_item-todo\\" data-v-971ec5b0=\\"\\">Complete online JavaScript course</div>
      </div>
      <div class=\\"delete_todo\\" data-v-971ec5b0=\\"\\"><img src=\\"/src/assets/images/icon-cross.svg\\" alt=\\"cross icon\\" data-v-971ec5b0=\\"\\"></div>
    </div>
    <div class=\\"todo_item\\" data-v-971ec5b0=\\"\\" data-v-4afa86a5=\\"\\">
      <div class=\\"todo_label\\" data-v-971ec5b0=\\"\\"><input type=\\"checkbox\\" class=\\"checkbox\\" data-v-971ec5b0=\\"\\">
        <div class=\\"todo_item-todo\\" data-v-971ec5b0=\\"\\">Jog around the park 3x</div>
      </div>
      <div class=\\"delete_todo\\" data-v-971ec5b0=\\"\\"><img src=\\"/src/assets/images/icon-cross.svg\\" alt=\\"cross icon\\" data-v-971ec5b0=\\"\\"></div>
    </div>
    <div class=\\"todo_item\\" data-v-971ec5b0=\\"\\" data-v-4afa86a5=\\"\\">
      <div class=\\"todo_label\\" data-v-971ec5b0=\\"\\"><input type=\\"checkbox\\" class=\\"checkbox\\" data-v-971ec5b0=\\"\\">
        <div class=\\"todo_item-todo\\" data-v-971ec5b0=\\"\\">10 minutes meditation</div>
      </div>
      <div class=\\"delete_todo\\" data-v-971ec5b0=\\"\\"><img src=\\"/src/assets/images/icon-cross.svg\\" alt=\\"cross icon\\" data-v-971ec5b0=\\"\\"></div>
    </div>
    <div class=\\"todo_item\\" data-v-971ec5b0=\\"\\" data-v-4afa86a5=\\"\\">
      <div class=\\"todo_label\\" data-v-971ec5b0=\\"\\"><input type=\\"checkbox\\" class=\\"checkbox\\" data-v-971ec5b0=\\"\\">
        <div class=\\"todo_item-todo\\" data-v-971ec5b0=\\"\\">Read for 1 hour</div>
      </div>
      <div class=\\"delete_todo\\" data-v-971ec5b0=\\"\\"><img src=\\"/src/assets/images/icon-cross.svg\\" alt=\\"cross icon\\" data-v-971ec5b0=\\"\\"></div>
    </div>
    <div class=\\"footer_container\\" data-v-4afa86a5=\\"\\">
      <div class=\\"items_left\\" data-v-4afa86a5=\\"\\">
        <div data-v-4afa86a5=\\"\\">4 items left</div>
      </div>
      <div class=\\"filters\\" data-v-4afa86a5=\\"\\">
        <p class=\\"active\\" data-v-4afa86a5=\\"\\">All</p>
        <p class=\\"\\" data-v-4afa86a5=\\"\\">Active</p>
        <p class=\\"\\" data-v-4afa86a5=\\"\\">Completed</p>
      </div>
      <div class=\\"clear_completed\\" data-v-4afa86a5=\\"\\"><button class=\\"clear_btn\\" data-v-4afa86a5=\\"\\">Clear Completed</button></div>
    </div>
  </div>
</div>"
`;
